type2.go

```diff
@@ -98,8 +102,49 @@ type typeParser struct {
 	parseUncommon        uncommonTypeParseFunc
 }
 
+func (p *typeParser) hasTag(off uint64) bool {
+	return p.typesData[off]&(1<<1) != 0
+}
+
 func (p *typeParser) resolveName(ptr uint64, flags uint8) (string, int) {
-	return resolveName(p.typesData, ptr, flags)
+	if GoVersionCompare(p.version, "go1.17beta1") < 0 {
+		// before go1.17, the length of name used fixed 2-byte encoding.
+		return resolveName(p.typesData, ptr, flags)
+	} else {
+		// go1.17, the length of name used variant encoding.
+		i, l := binary.Uvarint(p.typesData[ptr+1:])
+		name := string(p.typesData[ptr+1+uint64(l) : ptr+1+uint64(l)+i])
+		nl := int(i)
+
+		if flags&tflagExtraStar != 0 {
+			// typ.Name = strData[1:]
+			return name[1:], nl - 1
+		}
+		return name, nl
+	}
+}
+
+func (p *typeParser) resolveTag(o uint64) string {
+	var (
+		nl, nll = uint64(0), 2
+		tl, tll = uint64(0), 2
+	)
+	if !p.hasTag(o) {
+		return ""
+	}
+	if GoVersionCompare(p.version, "go1.17beta1") < 0 {
```

Comment:
If you check the master branch, I used similar logic within the parsing functions to check for different Go versions. This turned out to be very hard to maintain and resulted in code that was hard to read. Because of this, the type parser was redesigned. Instead it uses different parsing functions depending on the arch and Go version. See [this](https://github.com/goretk/gore/blob/600ad72ebd0503a7a25b673be27ae353568fb1c0/type2.go#L534) for an example. The [constructor](https://github.com/goretk/gore/blob/600ad72ebd0503a7a25b673be27ae353568fb1c0/type2.go#L21) selects the correct parsing function based on Go version and 32 vs 64.